core/src/Streamly/Internal/FileSystem/Path/Common.hs

```diff
@@ -0,0 +1,368 @@
+import Streamly.Internal.Data.Stream (Stream)
+import System.IO.Unsafe (unsafePerformIO)
+
+import qualified Streamly.Internal.Data.Array as Array
+import qualified Streamly.Internal.Data.Fold as Fold
+import qualified Streamly.Internal.Data.MutArray as MutArray
+import qualified Streamly.Internal.Data.Stream as Stream
+
+data OS = Windows | Posix
+
+------------------------------------------------------------------------------
+-- Construction
+------------------------------------------------------------------------------
+
+-- A chunk is essentially an untyped Array i.e. Array Word8.  We can either use
+-- the term ByteArray for that or just Chunk. The latter is shorter and we have
+-- been using it consistently in streamly. We use "bytes" for a stream of
+-- bytes.
+
+-- | /Unsafe/: The user is responsible to maintain the invariants mentioned in
+-- the definition of the 'Path' type. On Windows, the array passed must be a
+-- multiple of 2 bytes as the underlying representation uses 'Word16'.
+{-# INLINE unsafeFromChunk #-}
+unsafeFromChunk ::
+#ifdef DEBUG
+    Unbox a =>
+#endif
+    Array Word8 -> Array a
+unsafeFromChunk =
+#ifndef DEBUG
+    Array.castUnsafe
+#else
+    fromJust . fromChunk
+#endif
+
+-- XXX Also check for invalid chars on windows.
+
+-- | On Posix it may fail if the byte array contains null characters. On
+-- Windows the array passed must be a multiple of 2 bytes as the underlying
+-- representation uses 'Word16'.
+--
+-- Throws 'InvalidPath'.
+fromChunk :: (MonadThrow m, Unbox a) => Array Word8 -> m (Array a)
+fromChunk arr =
+    case Array.cast arr of
+        Nothing ->
+            -- XXX Windows only message.
+            throwM
+                $ InvalidPath
+                $ "Encoded path length " ++ show (Array.byteLength arr)
+                    ++ " is not a multiple of 16-bit."
+        Just x -> pure x
+
+-- | Convert 'Path' to an array of bytes.
+toChunk :: Array a -> Array Word8
+toChunk = Array.asBytes
+
+unsafeFromChars :: (Unbox a) =>
+       (Char -> Bool)
+    -> (Stream Identity Char -> Stream Identity a)
+    -> Stream Identity Char
+    -> Array a
+unsafeFromChars _p encode s =
+#ifndef DEBUG
+    let n = runIdentity $ Stream.fold Fold.length s
+     in Array.fromPureStreamN n (encode s)
+#else
+     fromJust (fromChars _p encode s)
+#endif
+
+-- XXX Sanitize the path - remove duplicate separators, . segments, trailing .
+-- XXX Writing a custom fold for parsing a Posix path may be better for
+-- efficient bulk parsing when needed. We need the same code to validate a
+-- Chunk where we do not need to create an array.
+fromChars :: (MonadThrow m, Unbox a) =>
+       (Char -> Bool)
+    -> (Stream Identity Char -> Stream Identity a)
+    -> Stream Identity Char
+    -> m (Array a)
+fromChars p encode s =
+    -- XXX on windows terminate at first invalid char
+    let lengths = Fold.tee Fold.length (Fold.takeEndBy_ p Fold.length)
+        (n, n1) = runIdentity $ Stream.fold lengths s
```

Comment:
`n == n1` iff the stream is completely consumed. Ie. no NULL char.
Is that right?

Reply:
yes.